### ISSUE TITLE
Pytorch lightning switched to cpu in the middle of training. How can I debug this?

### ISSUE THREAD

**Issue by samikhenissi** [PyTorchLightning/pytorch-lightning#2658]
## 🐛 Bug

![image](https://user-images.githubusercontent.com/20474484/88071206-8bfabe80-cb41-11ea-8043-c09c57ef00f8.png)

So I am training a model and suddenly PyTorch lightning started using the CPU. I checked the GPU status and it is working normally.
I have never seen this before. Can Pytorch start using CPU in the middle of training (at the next epoch)?
The weird thing is the next epoch was using the gpu.

EDIT: Some asked how I noticed it is not using GPU.
I noticed the difference by looking at the it/s displayed in the verbose. It dropped to 13 it/s from 130 in one epoch and then rebounded to 150 it/s. I immediately took a look at the CUDA usage and It was very low. 

### To Reproduce

This is the training loop: Pretty standard

`    model = MatrixFactorizationAttention(att_mf_config)

    train_loader = sample_generator.instance_a_train_loader_attention(att_mf_config['batch_size'])
    val_loader = sample_generator.instance_val_loader_attention
    # wandb_logger.log_hyperparams(att_mf_config)
    early_stop_callback = EarlyStopping(
        monitor='HR',
        min_delta=0.01,
        patience=10,
        verbose=True,
        mode='max'
    )

    trainer = pl.Trainer(gpus=1, max_epochs=150, precision=16,checkpoint_callback=False,logger = False,early_stop_callback=early_stop_callback)
    trainer.fit(model, train_dataloader=train_loader, val_dataloaders=val_loader)
`

EDIT: apart from this training loop, I am using a standard PyTorch model with a regular classification pipeline. 

### Environment

Win 10 with pytorch lightning running on pycharm
Cuda 10.2
GPU 2060 super
pytorch: nightly version

**Comment by rohitgr7** [comment]
Weird, mind share the code in colab or something?

**Comment by awaelchli** [comment]
How did you come to the conclusion that it is switching to CPU? The screenshot certainly does not give us this information.

**Comment by awaelchli** [comment]
You probably saw it in tensorboard? Is it one entire epoch?

**Comment by samikhenissi** [comment]
> How did you come to the conclusion that it is switching to CPU? The screenshot certainly does not give us this information.
In the verbose, you can see the number of iterations per second. It dropped to 13 it/s from 130 it/s. I also monitored the CUDA memory.

**Comment by awaelchli** [comment]
Can we test this with a callback that checks the device like so:

```python 
from pytorch_lightning import Callback

class DeviceCallback(Callback):

    def on_batch_start(self, trainer, pl_module):
        assert next(pl_module.parameters()).device.type == "cuda"

... 
trainer = Trainer(..., callbacks=[DeviceCallback()])
```
If this assertion does not fail and your iterations decrease, it might be a data loading issue.

**Comment by samikhenissi** [comment]
Thank you. It did not fail indeed when the speed decreased. Any idea what may cause this behavior or how to investigate the data loading issue? 
I create the trainloader from a dataframe using a standard method.
`    def instance_a_train_loader_attention(self,batch_size):
        """instance train loader for one training epoch"""
        users, items, ratings = self.train_ratings['userId'].tolist(), self.train_ratings['itemId'].tolist(), self.train_ratings['rating'].tolist()
        items_rated_byuser, users_rated_item = self.train_ratings['item_sample'].tolist(), self.train_ratings['user_sample'].tolist()
        dataset = UserItemRatingDatasetAttention(user_tensor=torch.LongTensor(users),
                                        item_tensor=torch.LongTensor(items),
                                        target_tensor=torch.FloatTensor(ratings),
                                        items_rated_user=torch.LongTensor(items_rated_byuser),
                                        users_rated_item_tensor=torch.LongTensor(users_rated_item))
        return DataLoader(dataset, batch_size=batch_size, shuffle=True)
`

**Comment by samikhenissi** [comment]
Anyone who wants to reproduce the issue here is the code: https://github.com/samikhenissi/pl_neumf

I am not sure if it is a hardware issue from my part or a bug. The speed just randomly drops for one iteration and goes back up again. The callback showed that I am still using the GPU for the model but with very low speed.

**Comment by awaelchli** [comment]
Can reproduce using this code, but don't know yet what's causing it. My observations so far:
- it also happens on cpu (if you set gpus=0)
- does not look to be related to logging, row_log_interval, log_save_interval, progress bar refresh rate
- does not look like a data loading issue
- in my case, the it/s display does not drop, I only see interruptions directly looking at the iteration counter.

**Comment by Borda** [comment]
I guess this was resolved, otherwise feel free to reopen... 🐰

**Comment by wadmes** [comment]
Similar case here, in the middle of training, after end of some epoch,  the model goes to CPU memory for some reason

```
[2090.87447] meta_data.device cuda:2
Epoch 17: 100%|##########| 175/175 [01:54<00:00,  1.52it/s, v_num=23, train_loss=1.710, imp_max_p=0.128, train_top3_accuracy=1.000, val_max_p=0.116, val_pred_max_p=0.388, val_correct_num=11.40, val_total_num=11.90]
[2096.24900] graph.region_meta_data.device cuda:4                       
[2096.24934] next(model.parameters()).device cpu
[2096.24948] meta_data.device cuda:4
[2096.73534] graph.region_meta_data.device cuda:5
[2096.73586] next(model.parameters()).device cpu
[2096.73615] meta_data.device cuda:5
[2096.86830] graph.region_meta_data.device cuda:1
[2096.86887] next(model.parameters()).device cpu
[2096.86914] meta_data.device cuda:1
```
